homeassistant/components/portainer/config_flow.py

```diff
@@ -0,0 +1,129 @@
+"""Config flow for the portainer integration."""
+
+from __future__ import annotations
+
+import logging
+from typing import Any
+
+from aiohttp import CookieJar
+from pyportainer import (
+    Portainer,
+    PortainerAuthenticationError,
+    PortainerConnectionError,
+    PortainerTimeoutError,
+)
+import voluptuous as vol
+from yarl import URL
+
+from homeassistant.config_entries import ConfigFlow, ConfigFlowResult
+from homeassistant.const import CONF_API_KEY, CONF_URL, CONF_VERIFY_SSL
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.aiohttp_client import async_create_clientsession
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_URL): str,
+        vol.Required(CONF_API_KEY): str,
+        vol.Required(CONF_VERIFY_SSL, default=True): bool,
+    }
+)
+
+
+async def _validate_input(hass: HomeAssistant, data: dict[str, Any]) -> dict[str, Any]:
+    """Validate the user input allows us to connect."""
+    url = URL(data[CONF_URL]).human_repr()
+    session = async_create_clientsession(
+        hass,
+        data[CONF_VERIFY_SSL],
+        cookie_jar=CookieJar(unsafe=True),
+    )
+    client = Portainer(
+        api_url=url,
+        api_key=data[CONF_API_KEY],
+        session=session,
+    )
+
```

Comment:
the url normalisation is not required at this point, see one of the next comments
```suggestion
    session = async_create_clientsession(
        hass,
        data[CONF_VERIFY_SSL],
        cookie_jar=CookieJar(unsafe=True),
    )
    client = Portainer(
        api_url=data[CONF_URL],
        api_key=data[CONF_API_KEY],
        session=session,
    )

```